File: bgpx/capture.py
"""BGP packet capture via tcpdump subprocess — results go to EventBus."""

import asyncio
import ipaddress
import logging
import re
import shutil
from typing import Optional

from bgpx.events import EventBus

log = logging.getLogger(__name__)

# BGP message types (first byte after the 16-byte marker + 2-byte length)
_BGP_MSG_TYPES = {1: "OPEN", 2: "UPDATE", 3: "NOTIFICATION", 4: "KEEPALIVE"}

# TCP flag abbreviations
_FLAG_MAP = {
    "S":  "SYN", "R": "RST", "F": "FIN",
    "P":  "PSH", ".": "ACK",
    "S.": "SYN-ACK", "F.": "FIN-ACK", "R.": "RST-ACK", "P.": "PSH-ACK",
}

# Parse a tcpdump text line (with -X hex output lines ignored here)
_PKT_RE = re.compile(
    r"(\d{2}:\d{2}:\d{2}\.\d+)"        # timestamp
    r".+?\s(In|Out)\s+IP\s"            # direction
    r"([\d.]+)\.(\d+)\s+>\s+"          # src ip.port
    r"([\d.]+)\.(\d+):\s+"             # dst ip.port
    r"Flags\s+\[([^\]]+)\]"            # TCP flags
    r".*?length\s+(\d+)"               # TCP payload length
)

# Hex line from tcpdump -X:  \t0x0000:  ffff ffff ...
_HEX_RE = re.compile(r"^\s+0x([0-9a-f]+):\s+(.+)$")


class PacketCapture:
    def __init__(self, events: EventBus):
        self._events = events
        self._proc:  Optional[asyncio.subprocess.Process] = None
        self._task:  Optional[asyncio.Task] = None
        self.running = False

    # ── Public ────────────────────────────────────────────────────────────────

    async def start(self, peer_ip: str) -> bool:
        if self.running:
            await self.stop()
        try:
            ipaddress.ip_address(peer_ip)
        except ValueError:
            self._emit("error", f"Invalid peer IP address: {peer_ip!r}")
            return False
        if not shutil.which("tcpdump"):
            self._emit("error", "tcpdump not found — install: apt install tcpdump")
            return False
        self.running = True
        self._task = asyncio.create_task(self._run(peer_ip))
        return True

    async def stop(self) -> None:
        self.running = False
        if self._proc:
            try:
                self._proc.terminate()
            except ProcessLookupError:
                pass
        if self._task:
            self._task.cancel()
            try:
                await self._task
            except (asyncio.CancelledError, Exception):
                pass
        self._proc = None
        self._task = None

    # ── Internal ──────────────────────────────────────────────────────────────

    async def _run(self, peer_ip: str) -> None:
        # -l  line-buffered
        # -nn no name resolution
        # -X  print hex payload (for BGP type detection)
        # -i any  all interfaces
        cmd = ["tcpdump", "-l", "-nn", "-X", "-i", "any",
               f"host {peer_ip} and port 179"]

        self._emit("info", "Capture started  →  " + " ".join(cmd))
        log.info("Capture: " + " ".join(cmd))

        try:
            self._proc = await asyncio.create_subprocess_exec(
                *cmd,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.STDOUT,
            )

            # We accumulate hex lines belonging to the current packet
            current_pkt: Optional[dict] = None
            hex_bytes: list[int] = []

            async for raw in self._proc.stdout:
                if not self.running:
                    break
                line = raw.decode(errors="replace").rstrip()
                if not line:
                    continue

                m_pkt = _PKT_RE.search(line)
                if m_pkt:
                    # Flush previous packet first
                    if current_pkt is not None:
                        self._flush(current_pkt, hex_bytes, peer_ip)

                    ts, direction, src_ip, src_port, dst_ip, dst_port, flags, length = m_pkt.groups()
                    current_pkt = dict(
                        ts=ts, direction=direction,
                        src=f"{src_ip}:{src_port}",
                        dst=f"{dst_ip}:{dst_port}",
                        src_ip=src_ip, dst_ip=dst_ip,
                        flags=flags, length=int(length),
                    )
                    hex_bytes = []
                    continue

                m_hex = _HEX_RE.match(line)
                if m_hex and current_pkt is not None:
                    raw_hex = m_hex.group(2).replace(" ", "")
                    try:
                        hex_bytes += [int(raw_hex[i:i+2], 16)
                                      for i in range(0, len(raw_hex), 2)]
                    except ValueError:
                        pass

            # Flush last packet
            if current_pkt is not None:
                self._flush(current_pkt, hex_bytes, peer_ip)

        except FileNotFoundError:
            self._emit("error", "tcpdump not found")
        except PermissionError:
            self._emit("error", "tcpdump needs root — run bgpx with sudo")
        except asyncio.CancelledError:
            raise
        except Exception as e:
            self._emit("error", f"Capture error: {e}", error=str(e))
        finally:
            if self._proc:
                try:
                    self._proc.terminate()
                    await self._proc.wait()
                except Exception:
                    pass
            self.running = False
            self._emit("info", "Capture stopped")
            log.info("Capture stopped")

    def _flush(self, pkt: dict, hex_bytes: list[int], peer_ip: str) -> None:
        """Emit one packet event, enriched with BGP type if detectable."""
        flags_label = _FLAG_MAP.get(pkt["flags"], pkt["flags"])
        direction   = "→" if pkt["dst_ip"] == peer_ip else "←"
        bgp_type    = _detect_bgp_type(hex_bytes) if hex_bytes else None
        tcp_event   = _tcp_event(pkt["flags"])

        # Build human-readable summary
        summary = (
            f"[{flags_label}]  "
            f"{pkt['src']}  {direction}  {pkt['dst']}"
            f"  len={pkt['length']}"
        )
        if bgp_type:
            summary += f"  ▶ BGP {bgp_type}"
        elif tcp_event:
            summary += f"  ({tcp_event})"

        self._events.emit(
            "capture", "packet", summary,
            src=pkt["src"], dst=pkt["dst"],
            direction=direction,
            flags=flags_label,
            length=pkt["length"],
            bgp_type=bgp_type,
            tcp_event=tcp_event,
        )

    def _emit(self, level: str, message: str, **kw) -> None:
        self._events.emit("capture", level, message, **kw)


# ── Helpers ───────────────────────────────────────────────────────────────────

def _detect_bgp_type(raw: list[int]) -> Optional[str]:
    """
    BGP header: 16 bytes marker (0xFF) + 2 bytes length + 1 byte type.
    In a TCP stream the BGP header starts at byte 0 of the payload.
    tcpdump's hex includes the IP+TCP headers before the payload, so
    we scan for the 0xFF marker sequence.
    """
    marker = [0xFF] * 16
    for i in range(len(raw) - 18):
        if raw[i:i + 16] == marker:
            msg_type = raw[i + 18] if i + 18 < len(raw) else None
            if msg_type in _BGP_MSG_TYPES:
                return _BGP_MSG_TYPES[msg_type]
    return None


def _tcp_event(flags: str) -> Optional[str]:
    if flags in ("S",):     return "TCP SYN — connection attempt"
    if flags in ("S.",):    return "TCP SYN-ACK — connection accepted"
    if flags in ("R", "R."): return "TCP RST — connection refused/reset"
    if flags in ("F", "F."): return "TCP FIN — connection closing"
    return None

File: src/capture.rs
use crate::app::{bad, ApiError, App, Shared, Task};
use axum::{extract::State, Json};
use regex::Regex;
use serde_json::{json, Value};
use std::process::Stdio;
use tokio::{
    io::{AsyncBufReadExt, BufReader},
    process::Command,
    sync::oneshot,
};

pub async fn start(State(app): State<Shared>) -> Result<Json<Value>, ApiError> {
    let mut control = app.control.lock().await;
    let peer = {
        let data = app.data.lock().unwrap();
        if !data.running {
            return Err(bad("No active session"));
        }
        data.config.as_ref().unwrap().peer_ip
    };
    App::stop_task(&mut control.capture).await;
    let mut child = Command::new("tcpdump")
        .args([
            "-l",
            "-nn",
            "-X",
            "-i",
            "any",
            &format!("host {peer} and port 179"),
        ])
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .kill_on_drop(true)
        .spawn()
        .map_err(bad)?;
    let stdout = child.stdout.take().unwrap();
    let stderr = child.stderr.take().unwrap();
    let (stop, mut rx) = oneshot::channel();
    app.data.lock().unwrap().capturing = true;
    app.emit(
        "capture",
        "info",
        "Capture started",
        json!({"running":true}),
    );
    let cloned = app.clone();
    control.capture = Some(Task {
        stop,
        join: tokio::spawn(async move {
            let app = cloned;
            let mut lines = BufReader::new(stdout).lines();
            let mut errors = BufReader::new(stderr).lines();
            let regex=Regex::new(r"\d{2}:\d{2}:\d{2}\.\d+.+?\s(In|Out)\s+IP\s([\d.]+)\.(\d+)\s+>\s+([\d.]+)\.(\d+):\s+Flags\s+\[([^\]]+)\].*?length\s+(\d+)").unwrap();
            let mut packet = None;
            let mut bytes = Vec::new();
            let mut stderr_open = true;
            loop {
                tokio::select! {
                    _=&mut rx=>break,
                    line=lines.next_line()=>match line{
                        Ok(Some(line))=>{
                            if let Some(c)=regex.captures(&line){
                                flush(&app,packet.take(),&bytes);bytes.clear();
                                let flags=match &c[6]{"S"=>"SYN","R"=>"RST","F"=>"FIN","P"=>"PSH","."=>"ACK","S."=>"SYN-ACK","F."=>"FIN-ACK","R."=>"RST-ACK","P."=>"PSH-ACK",f=>f};
                                let event=match &c[6]{"S"=>Some("TCP SYN - connection attempt"),"S."=>Some("TCP SYN-ACK - connection accepted"),"R"|"R."=>Some("TCP RST - connection refused/reset"),"F"|"F."=>Some("TCP FIN - connection closing"),_=>None};
                                packet=Some(json!({"src":format!("{}:{}",&c[2],&c[3]),"dst":format!("{}:{}",&c[4],&c[5]),"direction":if c[4]==peer.to_string(){"→"}else{"←"},"flags":flags,"length":c[7].parse::<u32>().unwrap_or(0),"tcp_event":event}));
                            }else if line.trim_start().starts_with("0x"){
                                if let Some((_,hex))=line.split_once(':'){
                                    for word in hex.split_whitespace().take(8){
                                        if ![2,4].contains(&word.len())||!word.bytes().all(|b|b.is_ascii_hexdigit()){break;}
                                        bytes.push(u8::from_str_radix(&word[..2],16).unwrap());
                                        if word.len()==4 {bytes.push(u8::from_str_radix(&word[2..],16).unwrap());}
                                    }
                                }
                            }
                        }
                        Ok(None)=>break,
                        Err(e)=>{app.emit("capture","error",e,json!({}));break;}
                    },
                    line=errors.next_line(),if stderr_open=>match line{
                        Ok(Some(line))=>app.emit("capture","info",line,json!({})),
                        _=>stderr_open=false,
                    }
                }
            }
            flush(&app, packet, &bytes);
            let _ = child.kill().await;
            let _ = child.wait().await;
            app.data.lock().unwrap().capturing = false;
            app.emit(
                "capture",
                "info",
                "Capture stopped",
                json!({"running":false}),
            );
        }),
    });
    Ok(Json(json!({"ok":true})))
}
fn flush(app: &Shared, packet: Option<Value>, bytes: &[u8]) {
    let Some(mut p) = packet else { return };
    let kind = bytes.windows(19).find_map(|w| {
        if w[..16] == [255; 16] {
            match w[18] {
                1 => Some("OPEN"),
                2 => Some("UPDATE"),
                3 => Some("NOTIFICATION"),
                4 => Some("KEEPALIVE"),
                _ => None,
            }
        } else {
            None
        }
    });
    p["bgp_type"] = json!(kind);
    let message = format!(
        "[{}] {} {} {} len={}{}",
        p["flags"].as_str().unwrap(),
        p["src"].as_str().unwrap(),
        p["direction"].as_str().unwrap(),
        p["dst"].as_str().unwrap(),
        p["length"],
        kind.map(|k| format!(" BGP {k}")).unwrap_or_default()
    );
    app.emit("capture", "packet", message, p);
}
pub async fn stop(State(app): State<Shared>) -> Json<Value> {
    App::stop_task(&mut app.control.lock().await.capture).await;
    Json(json!({"ok":true}))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn packet_event_contains_bgp_type() {
        let app = App::new(None);
        let packet = json!({"src":"192.0.2.1:179","dst":"192.0.2.2:40000","direction":"in","flags":"ACK","length":19,"tcp_event":null});
        let mut payload = vec![0; 40];
        payload.extend(crate::wire::message(4, &[]));
        flush(&app, Some(packet), &payload);
        let data = app.data.lock().unwrap();
        let event: Value = serde_json::from_str(data.history.back().unwrap()).unwrap();
        assert_eq!(event["type"], "capture");
        assert_eq!(event["bgp_type"], "KEEPALIVE");
        assert_eq!(event["src"], "192.0.2.1:179");
    }
}
